src/nano_contracts/methodArg.ts

```diff
@@ -0,0 +1,297 @@
+      .transform(val => Buffer.from(val, 'hex'))
+      .safeParse(inputVal);
+    if (!parse.success) {
+      ctx.addIssue({
+        code: z.ZodIssueCode.custom,
+        message: `Value is invalid ${type}: ${parse.error}`,
+        fatal: true,
+      });
+    } else {
+      return parse.data;
+    }
+  } else if (type === 'bool') {
+    const parse = z
+      .boolean()
+      .or(z.union([z.literal('true'), z.literal('false')]).transform(val => val === 'true'))
+      .safeParse(inputVal);
+    if (!parse.success) {
+      ctx.addIssue({
+        code: z.ZodIssueCode.custom,
+        message: `Value is invalid bool: ${parse.error}`,
+        fatal: true,
+      });
+    } else {
+      return parse.data;
+    }
+  } else if (type === 'str') {
+    const parse = z.string().safeParse(inputVal);
+    if (!parse.success) {
+      ctx.addIssue({
+        code: z.ZodIssueCode.custom,
+        message: `Value is invalid str: ${parse.error}`,
+        fatal: true,
+      });
+    } else {
+      return parse.data;
+    }
+  } else if (type === 'Address') {
+    const parse = z.string().safeParse(inputVal);
+    if (!parse.success) {
+      ctx.addIssue({
+        code: z.ZodIssueCode.custom,
+        message: `Value is invalid Address: ${parse.error}`,
+        fatal: true,
+      });
+    } else {
+      const address = new Address(parse.data);
+      try {
+        address.validateAddress({ skipNetwork: true });
+        return parse.data;
+      } catch (err) {
+        ctx.addIssue({
+          code: z.ZodIssueCode.custom,
+          message: `Value is invalid Address: ${err instanceof Error ? err.message : String(err)}`,
+          fatal: true,
+        });
+      }
+    }
+  } else {
+    // No known types match the given type
+    ctx.addIssue({
+      code: z.ZodIssueCode.custom,
+      message: `Type(${type}) is not supported as a 'single' type`,
+      fatal: true,
+    });
+  }
+
+  // Meant to keep the typing correct
+  return z.NEVER;
+}
+
+/**
+ * Type and value validation for non-container types.
+ * Returns the internal TS type for the argument given.
+ */
+const SingleValueApiInputScheme = z
+  .tuple([
+    z.string(), // type
+    NanoContractArgumentApiInputSchema, // value
+  ])
+  .transform((value, ctx) => {
+    return refineSingleValue(ctx, value[1], value[0]);
+  });
+
+/**
+ * Type and value validation for Optional types.
+ * Returns the internal TS type for the argument given.
+ */
+const OptionalApiInputScheme = z
+  .tuple([
+    z.string(), // Inner type
+    NanoContractArgumentApiInputSchema, // value
+  ])
+  .transform((value, ctx) => {
+    const parse = z.null().safeParse(value[1]);
+    if (parse.success) {
+      return parse.data;
+    }
+    // value is not null, should transform based on the type
+    return refineSingleValue(ctx, value[1], value[0]);
+  });
+
+/**
+ * Type and value validation for SignedData types.
+ * returns an instance of NanoContractSignedData
+ */
+const SignedDataApiInputScheme = z
```

Comment:
This is meant to receive the signed data string `<signature>,<value>,<type>`, validate that `value` is of type `type`, `signature` is a hex string and transform to the internal type `NanoContractSignedData`